File: mobile/modules/neptranslate-ocr/ios/NeptranslateOcrModule.swift
import ExpoModulesCore
import MLKitTextRecognition
import MLKitTextRecognitionDevanagari
import MLKitVision

public class NeptranslateOcrModule: Module {
  public func definition() -> ModuleDefinition {
    Name("NeptranslateOcr")

    AsyncFunction("recognize") { (uri: String, promise: Promise) in
      guard let url = URL(string: uri),
            let data = try? Data(contentsOf: url),
            let image = UIImage(data: data) else {
        promise.reject("empty", "The capture had no readable image.")
        return
      }
      let vision = VisionImage(image: image)
      vision.orientation = image.imageOrientation
      let latin = TextRecognizer.textRecognizer()
      let devanagari = TextRecognizer.textRecognizer(options: DevanagariTextRecognizerOptions())
      latin.process(vision) { latinResult, latinError in
        if let latinError {
          promise.reject("ocr", latinError.localizedDescription)
          return
        }
        devanagari.process(vision) { devaResult, devaError in
          if let devaError {
            promise.reject("ocr", devaError.localizedDescription)
            return
          }
          let blocks = Self.blocks(from: latinResult) + Self.blocks(from: devaResult, language: "ne")
          promise.resolve([
            "width": image.size.width,
            "height": image.size.height,
            "rotation": 0,
            "blocks": blocks,
          ])
        }
      }
    }
  }

  private static func blocks(from result: Text?, language: String = "en") -> [[String: Any]] {
    guard let result else { return [] }
    return result.blocks.map { block in
      [
        "text": block.text,
        "language": language,
        "confidence": 0.9,
        "frame": frame(block.frame),
        "cornerPoints": points(block.cornerPoints),
        "lines": block.lines.map { line in
          [
            "text": line.text,
            "confidence": 0.9,
            "frame": frame(line.frame),
            "cornerPoints": points(line.cornerPoints),
          ]
        },
      ]
    }
  }

  private static func frame(_ rect: CGRect) -> [String: Double] {
    [
      "x": rect.origin.x,
      "y": rect.origin.y,
      "width": rect.size.width,
      "height": rect.size.height,
    ]
  }

  private static func points(_ points: [NSValue]) -> [[String: Double]] {
    points.map { value in
      let point = value.cgPointValue
      return ["x": point.x, "y": point.y]
    }
  }
}

File: mobile/modules/neptranslate-ocr/android/src/main/java/expo/modules/neptranslateocr/NeptranslateOcrModule.kt
package expo.modules.neptranslateocr

import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import com.google.mlkit.vision.common.InputImage
import com.google.mlkit.vision.text.TextRecognition
import com.google.mlkit.vision.text.latin.TextRecognizerOptions
import com.google.mlkit.vision.text.devanagari.DevanagariTextRecognizerOptions
import android.net.Uri

class NeptranslateOcrModule : Module() {
  override fun definition() = ModuleDefinition {
    Name("NeptranslateOcr")

    AsyncFunction("recognize") { uri: String ->
      val context = appContext.reactContext ?: throw Exception("Camera OCR is unavailable.")
      val image = InputImage.fromFilePath(context, Uri.parse(uri))
      val latin = TextRecognition.getClient(TextRecognizerOptions.DEFAULT_OPTIONS)
      val deva = TextRecognition.getClient(DevanagariTextRecognizerOptions.Builder().build())
      val latinText = com.google.android.gms.tasks.Tasks.await(latin.process(image))
      val devaText = com.google.android.gms.tasks.Tasks.await(deva.process(image))
      mapOf(
        "width" to image.width,
        "height" to image.height,
        "rotation" to image.rotationDegrees,
        "blocks" to (blocks(latinText, "en") + blocks(devaText, "ne"))
      )
    }
  }

  private fun blocks(text: com.google.mlkit.vision.text.Text, language: String): List<Map<String, Any>> {
    return text.textBlocks.map { block ->
      mapOf(
        "text" to block.text,
        "language" to language,
        "confidence" to 0.9,
        "frame" to frame(block.boundingBox),
        "cornerPoints" to points(block.cornerPoints),
        "lines" to block.lines.map { line ->
          mapOf(
            "text" to line.text,
            "confidence" to 0.9,
            "frame" to frame(line.boundingBox),
            "cornerPoints" to points(line.cornerPoints)
          )
        }
      )
    }
  }

  private fun frame(box: android.graphics.Rect?): Map<String, Int> {
    val rect = box ?: android.graphics.Rect()
    return mapOf(
      "x" to rect.left,
      "y" to rect.top,
      "width" to rect.width(),
      "height" to rect.height()
    )
  }

  private fun points(points: Array<android.graphics.Point>?): List<Map<String, Int>> {
    return (points ?: emptyArray()).map { mapOf("x" to it.x, "y" to it.y) }
  }
}
